File: prep/conversion.cpp
#include <bits/stdc++.h>
#include <cstddef>
#include <cstring>
using namespace std;

using Fact = tuple<string, double, string>;
using Query = tuple<double,string,string>;
using Graph = unordered_map<string,vector<pair<string,double>>>;

Graph compile_facts(const vector<Fact>& facts) {
    Graph result;
    for (const auto& f : facts) {
        result[get<0>(f)].push_back(make_pair(get<2>(f),get<1>(f)));
        result[get<2>(f)].push_back(make_pair(get<0>(f),1./get<1>(f)));
    }
    return result;
}

double query_(Query q, Graph G, bool& succ, unordered_set<string> history) {
    double amount = get<0>(q);
    string from = get<1>(q);
    string to = get<2>(q);
    if (from == to) {
        succ = true;
        return amount;
    }
    if (history.find(from) != history.end()) {
        succ = false;
        return -1;
    }
    history.insert(from);
    for (const auto& n : G[from]) {
        double res = query_(make_tuple(amount*n.second,n.first,to), G, succ, history);
        if (succ) {
            return res;
        }
    }
    succ = false;
    return -1;
}    

double query(Query q, Graph G, bool& succ) {
    unordered_set<string> history;
    return query_(q, G, succ, history);
}

int main()
{
    vector<Fact> facts{{"m",3.28,"ft"},
                       {"ft",12,"in"},
                       {"hr",60,"min"},
                       {"min",60,"sec"}};

    vector<Query> queries{{2,"m","in"},
                          {13,"in","m"},
                          {13,"in","hr"}};

    Graph G = compile_facts(facts);
    
    for (const auto& q : queries) {
        bool success=true;
        double res = query(q, G, success);
        if (!success) {
            cout << get<0>(q) << get<1>(q) << " conversion to " << get<2>(q) << " not supported" << endl;
        } else {
            cout << get<0>(q) << get<1>(q) << " = " << res << get<2>(q) << endl;
        }
    }

    return 0;
}

File: prep/conversion.py
facts = [('m',3.28,'ft'),
         ('ft',12,'in'),
         ('hr',60,'min'),
         ('min',60,'sec')]

queries = [(2,'m','in'),
           (13,'in','m'),
           (13,'in','hr')]

from collections import defaultdict

def compile_facts(facts):
    G = defaultdict(list)
    for from_unit, factor, to_unit in facts:
        G[from_unit].append((to_unit,factor))
        G[to_unit].append((from_unit,1/factor))
    return G

FactGraph = compile_facts(facts)
print(FactGraph)

def query_(q, G,History):    
    amount, from_unit, to_unit = q
    if from_unit == to_unit:
        return amount
    successors = G[from_unit]
    for unit, factor in successors:
        if unit not in History:
            History.add(unit)
            res = query_((amount*factor, unit, to_unit),G,History)
            if res is not None:
                return res
    return None

def query(q, G):
    res = query_(q,G,set([]))
    if res is not None:
        return res
    else:
        return "not supported"

for q in queries:
    print(q,query(q,FactGraph))